### PR TITLE
Math: Library: Add the hifi4 exponential function

### PR DESCRIPTION
**Pull request by ShriramShastry**
The 32-bit HiFi4 exponential library function has an accuracy of 1e-4, a unit in last place error of 5.60032793, and output ranges from 0.0067379470 to 148.4131591026 for inputs from -5 to +5 (Q4.28) (Q9.23).

Signed-off-by: ShriramShastry <malladi.sastry@intel.com>

image

### CODE REVIEW

**Review by lgirdwood — COMMENTED** [review]
@andrula-song @singalsu pls review

File: src/math/Kconfig [hunk 1]
```diff
@@ -38,6 +38,15 @@ config SQRT_FIXED
 	  to  calculate square root.square function having positive number
 	  y as input and return the positive number x multiplied by itself (squared)
 
+config EXP_FIXED
```
[the remark this answers is not in the document]
Comment (andrula-song) [reply src/math/Kconfig:41]: actually there is a c version in PR https://github.com/thesofproject/sof/pull/6655, and in exp_fcn_hifi4.h will determine which version be used.

File: src/math/exp_fcn_hifi4.c [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+#include <sof/common.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <stddef.h>
+
+#ifdef EXPONENTIAL_HIFI4
+#include <xtensa/tie/xt_hifi4.h>
+
+
+static void mul_s64(ae_int64 in_0, ae_int64 in_1, ae_int64 * ptroutbitshi,
```
Comment (singalsu) [inline src/math/exp_fcn_hifi4.c:19]: Extra blanks after ae_int64 *.  You could also avoid this function proto with move of mul_s64() here.

File: src/math/exp_fcn_hifi4.c [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+	ae_int64 result;
+	ae_int64 u64_chi;
+	ae_int64 u64_clo;
+
+	mul_s64(a, b, &u64_chi, &u64_clo);
+
+	ae_int64 roundup = AE_AND64(u64_clo, BIT_MASK_LOW_Q27P5);
+
+	roundup = AE_SRLI64(roundup, 27);
+	ae_int64 temp = AE_OR64(AE_SLAI64(u64_chi, 36), AE_SRLI64(u64_clo, 28));
```
Comment (singalsu) [inline src/math/exp_fcn_hifi4.c:33]: declare variables in the begin of functions

File: src/math/exp_fcn_hifi4.c [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+	u64_clo = AE_ADD64(temp, roundup);
```
Comment (singalsu) [inline src/math/exp_fcn_hifi4.c:35]: `return AE_ADD64(temp, roundup);`

File: src/math/exp_fcn_hifi4.c [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+ * Return Type	: void
+ */
+static void mul_s64(ae_int64 in_0, ae_int64 in_1, ae_int64 *ptroutbitshi, ae_int64 *ptroutbitslo)
+{
+	ae_int64 producthihi, producthilo, productlolo;
+
+	ae_int32x2 in0_32 = AE_MOVINT32X2_FROMINT64(in_0);
+	ae_int32x2 in1_32 = AE_MOVINT32X2_FROMINT64(in_1);
+
+	ae_ep ep_lolo = AE_MOVEA(0);
```
Comment (singalsu) [inline src/math/exp_fcn_hifi4.c:56]: This seems to make it not compatible with hifi3. Did yo consider making such version?

File: src/math/exp_fcn_hifi4.c [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+	ae_ep ep_hilo = AE_MOVEA(0);
+	ae_ep ep_HL_LH = AE_MOVEA(0);
+
+	producthihi = AE_MUL32_HH(in0_32, in1_32);
+	AE_MULZAAD32USEP_HL_LH(ep_hilo, producthilo, in1_32, in0_32);
+	productlolo = AE_MUL32U_LL(in0_32, in1_32);
+
+	ae_int64 product_HL_LH_H = AE_SRAI72(ep_hilo, producthilo, 32);
```
Comment (singalsu) [inline src/math/exp_fcn_hifi4.c:64]: declare variables in the begin of function

File: src/math/exp_fcn_hifi4.c [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+	carry = AE_SRLI64(carry, 32);
+	ae_int64 producthi = AE_ADD64(producthihi, carry);
+
+	producthi = AE_ADD64(producthi, product_HL_LH_H);
+
+	*ptroutbitslo = productlolo;
+	*ptroutbitshi = producthi;
+}
+
+int64_t onebyfact_Q63[19] = {
```
Comment (singalsu) [inline src/math/exp_fcn_hifi4.c:80]: If this table is the same as for C version make it common, looks like it is?

File: src/math/exp_fcn_hifi4.c [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+	ae_int64 outlo;
+	ae_int64 qt;
+	ae_int64 onebyfact;
+
+	int64_t qt_temp;
+	int64_t b_n;
+
+	xtbool flag;
+
+	ae_int64 *iv = (ae_int64 *)&onebyfact_Q63[0];
```
Comment (singalsu) [inline src/math/exp_fcn_hifi4.c:129]: Declare all variables in the beginning

File: src/math/exp_fcn_hifi4.c [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+		mp = mul_s64_sr_sat_near(mp, (int64_t)x);
+
+		sign = AE_NEG64(qt);
+
+		flag = AE_LT64(qt, 0);
+		AE_MOVT64(qt, sign, flag);
+
+		qt_temp = qt;
+		if (qt_temp <= CONVERG_ERROR)
```
Comment (singalsu) [inline src/math/exp_fcn_hifi4.c:163]: Is there code cycles difference for less than vs. less or equal?

File: src/math/exp_fcn_hifi4.c [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+		ts = AE_ADD64(ts, temp);
+
+		mp = mul_s64_sr_sat_near(mp, (int64_t)x);
+
+		sign = AE_NEG64(qt);
+
+		flag = AE_LT64(qt, 0);
+		AE_MOVT64(qt, sign, flag);
+
+		qt_temp = qt;
```
Comment (singalsu) [inline src/math/exp_fcn_hifi4.c:162]: Can you compare qt directly to integer constant without need for qt_temp?

File: test/cmocka/src/math/arithmetic/exponential_hifi4.c [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+#include <string.h>
+
+#include <sof/math/exp_fcn_hifi4.h>
+#include <sof/audio/format.h>
+#include <rtos/string.h>
+#include <sof/common.h>
+
+#define ULP_TOLERANCE 5.60032793
+#define ULP_SCALE 0.0000999999999749
+/* testvector in Q4.28, -5 to +5 */
```
Comment (singalsu) [inline test/cmocka/src/math/arithmetic/exponential_hifi4.c:23]: This seems to be common with C version. If so,  use common header file instead.

**Review by ShriramShastry — COMMENTED** [review]
Thank you for looking at the code.
I incorporated the proposal and expressed my opinion.

**Review by btian1 — APPROVED** [review]

**Comment by ShriramShastry** [comment]
Because there is no HiFi4 (XCHAL_HAVE_HIFI4) supported platform, the **Unit test** failed the sample test vectors.

**Comment by singalsu** [comment]
> Because there is no HiFi4 (XCHAL_HAVE_HIFI4) supported platform, the **Unit test** failed the sample test vectors.

Would a HiFi3 version be possible to see the MCPS impact to DRC?

File: src/include/sof/math/exp_fcn_hifi4.h [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+#undef EXPONENTIAL_GENERIC
+#define EXPONENTIAL_HIFI4
+#endif
+
+#endif
+
+/* define constants */
+#define BIT_MASK_LOW_Q27P5 0x0000000008000000
+#define BIT_MASK_Q62P2 0x4000000000000000
+#define CONVERG_ERROR 28823037624320LL	// error smaller than 1e-4,1/2 ^ -44.7122876209085
```
Comment (singalsu) [inline src/include/sof/math/exp_fcn_hifi4.h:28]: Use C /* */ comments

File: src/include/sof/math/exp_fcn_hifi4.h [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+#define QUOTIENT_SCALE 0x400000000
+#define TERMS_Q23P9 0x800000
+#define LSHIFT_BITS 0x2000
```
Comment (singalsu) [inline src/include/sof/math/exp_fcn_hifi4.h:31]: I'm not sure if macro BIT() works for 64 bits but at least for 32 bits value if the macro definiton is bitfield nature you could use it to make the code more clear. E:g. Here BIT(13).

File: src/math/exp_fcn_hifi4.c (hunk 7 above)
Comment (singalsu) [reply src/math/exp_fcn_hifi4.c:124]: OK. you will need to get the C version #6655 merged first.

File: src/math/exp_fcn_hifi4.c [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+	ae_int64 temp;
+
+	ae_int64 *ponebyfact_Q63 = (ae_int64 *)&onebyfact_Q63[0];
+	ae_int64 ts = TERMS_Q23P9;
+	xtbool flag;
+
+	int64_t qt_temp;
+	int64_t b_n;
+
+	mp = (x + LSHIFT_BITS) >> 14; //x in Q50.14
```
Comment (singalsu) [inline src/math/exp_fcn_hifi4.c:177]: Use /* */ comment

File: src/math/exp_fcn_hifi4.c [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+
+	for (b_n = 0; b_n < ARRAY_SIZE(onebyfact_Q63); b_n++) {
+		ae_int64 sign;
+
+		AE_L64_IP(onebyfact, ponebyfact_Q63, 8);
+
+		mul_s64(mp, onebyfact, &outhi, &outlo);
+		qt = AE_OR64(AE_SLAI64(outhi, 45), AE_SRLI64(outlo, 19));
+
+		temp = AE_SRAI64(AE_ADD64(qt, QUOTIENT_SCALE), 35);
```
Comment (singalsu) [inline src/math/exp_fcn_hifi4.c:196]: You could use here a shift with round instruction instead separate add (of 0.5) before right shift.

File: test/cmocka/src/math/arithmetic/exponential_hifi4.c (hunk 11 above)
Comment (singalsu) [reply test/cmocka/src/math/arithmetic/exponential_hifi4.c:23]: Yep, also this needs C version merge.

**Review by singalsu — CHANGES_REQUESTED** [review]

File: src/math/Kconfig (hunk 1 above)
[the remark this answers is not in the document]
Comment (cujomalainey) [reply src/math/Kconfig:41]: but that will still break the build system if it doesnt support hifi no?

File: src/math/exp_fcn_hifi4.c [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+
+	ae_int32x2 in0_32 = AE_MOVINT32X2_FROMINT64(in_0);
+	ae_int32x2 in1_32 = AE_MOVINT32X2_FROMINT64(in_1);
+
+	ae_ep ep_lolo = AE_MOVEA(0);
+	ae_ep ep_hilo = AE_MOVEA(0);
+	ae_ep ep_HL_LH = AE_MOVEA(0);
+
+	producthihi = AE_MUL32_HH(in0_32, in1_32);
+	AE_MULZAAD32USEP_HL_LH(ep_hilo, producthilo, in1_32, in0_32);
```
Comment (cujomalainey) [inline src/math/exp_fcn_hifi4.c:83]: do we not care about the result? or is this one of those times where hifi looks like its passing by value but its actually a reference

File: src/math/exp_fcn_hifi4.c [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+ *	(	 P1H)P1L:000
+ *	(	 P2H)P2L:000
+ *	(P3H:	 P3L)000:000
+ *	--------------------
+ *	(ccc:bbb)aaa:P0L
+ * P1H, P2H, and P3H:P3L are added to the carry CRY0 + CRY1.This increase will
+ * not result in an overflow.
+ *
+ */
+static void mul_s64(ae_int64 in_0, ae_int64 in_1, ae_int64 *ptroutbitshi, ae_int64 *ptroutbitslo)
```
Comment (cujomalainey) [inline src/math/exp_fcn_hifi4.c:70]: this seems best suited to be in a generic math lib

File: src/math/exp_fcn_hifi4.c (hunk 14 above)
Comment (cujomalainey) [reply src/math/exp_fcn_hifi4.c:177]: you can also add this to the variable declaration

**Review by cujomalainey — CHANGES_REQUESTED** [review]

File: src/math/exp_fcn_hifi4.c (hunk 16 above)
Comment (lgirdwood) [reply src/math/exp_fcn_hifi4.c:83]: @cujomalainey fwiw, a lot of the Hifi intrinsics confuse static analysis tooling. This could be one of them.

**Comment by lgirdwood** [comment]
@ShriramShastry conflicts

File: src/math/exp_fcn_hifi4.c (hunk 16 above)
Comment (cujomalainey) [reply src/math/exp_fcn_hifi4.c:83]: That does surprise me after listening to Maciej complain about it.

Comment (ShriramShastry) [reply src/math/exp_fcn_hifi4.c:83]: If you could provide me some background information, it would be useful for me to be able to make the necessary changes if there is consensus and clear guidance.

Thank you very much

Comment (cujomalainey) [reply src/math/exp_fcn_hifi4.c:83]: We are trying to understand where the output of this function is. From what it looks like in the code everything is pass by value so there "should" be no effect on the variables, but we also know the hifi spec violates this part of C standard, so we want to make sure this line is doing as intended and gets a comment explaining that

Comment (ShriramShastry) [reply src/math/exp_fcn_hifi4.c:83]: So, what is the proposal here for me? 
Thank you

Comment (singalsu) [reply src/math/exp_fcn_hifi4.c:83]: Please adds comments what's the input for AE_MULZAAD32USEP_HL_LH() and what's the output.

**Comment by ShriramShastry** [comment]
> @ShriramShastry conflicts

addressed the commits conflicts in [https://github.com/thesofproject/sof/pull/7033](url)
Please review the pull request. 
Thank you

File: src/math/exp_fcn_hifi4.c [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+	/* Input-32x32-bit(in1_32xin0_32)into 72-bit multiplication operations */
+	/* Output-lower 64 bits of the result are stored in producthilo */
+	/* Output-upper eight bits are stored in ep_hilo */
+	AE_MULZAAD32USEP_HL_LH(ep_hilo, producthilo, in1_32, in0_32);
+	productlolo = AE_MUL32U_LL(in0_32, in1_32);
+
+	product_hl_lh_h = AE_SRAI72(ep_hilo, producthilo, 32);
+	product_hl_lh_l = AE_SLAI64(producthilo, 32);
+
+	AE_ADD72(ep_lolo, productlolo, ep_HL_LH, product_hl_lh_l);
```
Comment (cujomalainey) [inline src/math/exp_fcn_hifi4.c:101]: @ShriramShastry can we do the same here that we did for `AE_MULZAAD32USEP_HL_LH`? Unfortunetly this is hifi being bad that requires this

Comment (ShriramShastry) [reply src/math/exp_fcn_hifi4.c:101]: `AE_MULZAAD32USEP_HL_LH` is an intrinsic multiply add, whereas AE_ADD72 is a bit increment. 

The `AE_MULZAAD32USEP_HL_LH` is a `32x32`-bit unsigned times signed integer dual MAC with a result of `72` bits and `no saturation`.

`AE_ADD72` is a 72-bit increment of an `AE_EP`, `AE_DR` register pair by another 72-bit register pair.

Could you perhaps provide a numerical example to back up your abd opinion `AE_ADD72`?

Thank you

Comment (ShriramShastry) [reply src/math/exp_fcn_hifi4.c:101]: @cujomalainey , any feedback here ? Thanks

File: src/include/sof/math/exp_fcn_hifi4.h [hunk 19]
```diff
@@ -0,0 +1,30 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2022 Intel Corporation. All rights reserved.
+ *
+ * Author: Shriram Shastry <malladi.sastry@linux.intel.com>
+ *
+ */
+#ifndef __SOF_MATH_EXPONENTIAL_HIFI4_H__
```
Comment (singalsu) [inline src/include/sof/math/exp_fcn_hifi4.h:8]: This is usually following header name, so __SOF_MATH_EXP_FCN_HIFI4.H would be better.

File: src/include/sof/math/exp_fcn_hifi4.h [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+#include <stddef.h>
+
+#if defined(__XCC__)
+/* HiFi */
+#include <xtensa/config/core-isa.h>
+#if XCHAL_HAVE_HIFI4 == 1
+#define EXPONENTIAL_HIFI4 1
+
+#else
+#error "No XCHAL_HAVE_HIFIx found. Cannot build XCC EXP - HiFi module."
```
Comment (singalsu) [inline src/include/sof/math/exp_fcn_hifi4.h:21]: #error causes build with TGL HiFi3 to fail. I think the EXPONENTIAL_GENERIC should be set here to build generic version with HiFi3 compiler.

File: src/include/sof/math/exp_fcn_hifi4.h [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+#define __SOF_MATH_EXPONENTIAL_HIFI4_H__
+
+#include <stdint.h>
+#include <stddef.h>
+
+#if defined(__XCC__)
+/* HiFi */
+#include <xtensa/config/core-isa.h>
+#if XCHAL_HAVE_HIFI4 == 1
+#define EXPONENTIAL_HIFI4 1
```
Comment (singalsu) [inline src/include/sof/math/exp_fcn_hifi4.h:18]: prefix macros with SOFM_

File: src/math/CMakeLists.txt [hunk 22]
```diff
@@ -14,8 +14,8 @@ if(CONFIG_SQRT_FIXED)
         add_local_sources(sof sqrt_int16.c)
 endif()
 
-if(CONFIG_EXP_FIXED)
-        add_local_sources(sof exp_fcn.c)
+if(CONFIG_MATH_EXP)
```
Comment (singalsu) [inline src/math/CMakeLists.txt:17]: If this is changed, should also the Kconfig name change?

File: test/cmocka/src/math/arithmetic/exponential.c [hunk 23]
```diff
@@ -65,7 +66,7 @@ static void test_math_arithmetic_exponential_fixed(void **state)
 	int32_t b_i;
 
 	srand((unsigned int)time(NULL));
-	for (i = 0; i < 256; i++) {
+	for (i = 0; i < 1024; i++) {
```
Comment (singalsu) [inline test/cmocka/src/math/arithmetic/exponential.c:69]: This unrelated change could be in another commit with title indicating improvement of cmocka test coverage. I assume the old test with 256 points would work with hifi4 version too.

**Review by singalsu — CHANGES_REQUESTED** [review]

File: test/cmocka/src/math/arithmetic/exponential.c [hunk 24]
```diff
@@ -16,6 +16,7 @@
 #include <time.h>
 #include <stdlib.h>
 
+#include <sof/math/exp_fcn_hifi4.h>
```
Comment (singalsu) [inline test/cmocka/src/math/arithmetic/exponential.c:19]: Can include of this hifi4 header be avoided in applications? Just `#include <sof/math/exp_fcn.h>` should be enough. This way there would be need to #include for every xtensa hifi version?

Comment (ShriramShastry) [reply test/cmocka/src/math/arithmetic/exponential.c]: Yes, I can remove the HiFi4 header file and include every version of xtensa HiFi.

File: src/math/CMakeLists.txt (hunk 22 above)
Comment (ShriramShastry) [reply src/math/CMakeLists.txt:17]: `Kconfig` is not using CONFIG_MATH_EXP but `EXP_FIXED` and default its set to `n`

File: src/math/exp_fcn_hifi4.c (hunk 18 above)
Comment (cujomalainey) [reply src/math/exp_fcn_hifi4.c:99]: @ShriramShastry im just asking for comment (in the code) explaining whats going on with the variables since HiFi violates C standard here so its hard to tell what is the input and what is the output

**Review by singalsu — APPROVED** [review]

**Review by kv2019i — COMMENTED** [review]
Code looks ready to go, but the Kconfig addition/rename needs clarification. See comment inline.

**Comment by ShriramShastry** [comment]
> Code looks ready to go, but the Kconfig addition/rename needs clarification. See comment inline.

Kconfig, is enabled or disabled by below code, In the moment, it is disabled, to enable make default `n` to `y`

config EXP_FIXED
	bool "Exponential functions"
	default `n`
	help
	  By selecting this, the 32-bit sofm_exp_int32() function can be used to calculate
	  exponential values. With a maximum ulp of 5, an exponential function with
	  an input range of -5 to +5 gives positive numbers between 0.00673794699908547 and
	  148.413159102577. The precision of this function is 1e-4.

**Comment by ShriramShastry** [comment]
> > Code looks ready to go, but the Kconfig addition/rename needs clarification. See comment inline.
> 
> Kconfig, is enabled or disabled by below code, In the moment, it is disabled, to enable make default `n` to `y`
> 
> config EXP_FIXED bool "Exponential functions" default `n` help By selecting this, the 32-bit sofm_exp_int32() function can be used to calculate exponential values. With a maximum ulp of 5, an exponential function with an input range of -5 to +5 gives positive numbers between 0.00673794699908547 and 148.413159102577. The precision of this function is 1e-4.

Done !! Renamed `EXP_FIXED` to `MATH_EXP`.
Thanks

**Review by kv2019i — APPROVED** [review]
Thanks @ShriramShastry  , looks good now.

**Comment by ShriramShastry** [comment]
> Thanks @ShriramShastry , looks good now.

Thank you. Is there any other unresolved issue that needs to be addressed?

**Comment by kv2019i** [comment]
@cujomalainey Ok to proceed with this? You still have a requested-changes pending.

**Comment by cujomalainey** [comment]
> @cujomalainey Ok to proceed with this? You still have a requested-changes pending.

I'll re-review today

File: src/math/exp_fcn_hifi4.c (hunk 18 above)
Comment (cujomalainey) [reply src/math/exp_fcn_hifi4.c:99]: @ShriramShastry I don't see the comment for `AE_ADD72`, did you push?

**Comment by cujomalainey** [comment]
looks like my comment is still open, would be nice to get it fixed but i won't block on it

**Review by cujomalainey — APPROVED** [review]
